xdsl/dialects/experimental/dmp.py

```diff
@@ -0,0 +1,455 @@
+
+    Where `x` signifies the area that should be received,
+    and `o` the area that should be read from.
+
+    This data will be exchanged with the node of rank (my_rank -1)
+    """
+
+    name = "dmp.exchange_decl"
+
+    offset_: ParameterDef[builtin.DenseArrayBase]
+    size_: ParameterDef[builtin.DenseArrayBase]
+    source_offset_: ParameterDef[builtin.DenseArrayBase]
+    neighbor_: ParameterDef[builtin.DenseArrayBase]
+
+    def __init__(
+        self,
+        offset: Sequence[int],
+        size: Sequence[int],
+        source_offset: Sequence[int],
+        neighbor: Sequence[int],
+    ):
+        typ = builtin.i64
+        super().__init__(
+            [
+                builtin.DenseArrayBase.from_list(typ, offset),
+                builtin.DenseArrayBase.from_list(typ, size),
+                builtin.DenseArrayBase.from_list(typ, source_offset),
+                builtin.DenseArrayBase.from_list(typ, neighbor),
+            ]
+        )
+
+    @property
+    def offset(self) -> tuple[int, ...]:
+        data = self.offset_.as_tuple()
+        assert isa(data, tuple[int, ...])
+        return data
+
+    @property
+    def size(self) -> tuple[int, ...]:
+        data = self.size_.as_tuple()
+        assert isa(data, tuple[int, ...])
+        return data
+
+    @property
+    def source_offset(self) -> tuple[int, ...]:
+        data = self.source_offset_.as_tuple()
+        assert isa(data, tuple[int, ...])
+        return data
+
+    @property
+    def neighbor(self) -> tuple[int, ...]:
+        data = self.neighbor_.as_tuple()
+        assert isa(data, tuple[int, ...])
+        return data
+
+    @property
+    def elem_count(self) -> int:
+        return prod(self.size)
+
+    @property
+    def dim(self) -> int:
+        return len(self.size)
+
+    def source_area(self) -> HaloExchangeDecl:
+        """
+        Since a HaloExchangeDef by default specifies the area to receive into,
+        this method returns the area that should be read from.
+        """
+        # we set source_offset to all zero, so that repeated calls to source_area never
+        # return the dest area
+        return HaloExchangeDecl(
+            offset=tuple(
+                val + offs for val, offs in zip(self.offset, self.source_offset)
+            ),
+            size=self.size,
+            source_offset=tuple(0 for _ in range(len(self.source_offset))),
+            neighbor=self.neighbor,
+        )
+
+    def print_parameters(self, printer: Printer) -> None:
+        printer.print_string("<at [")
+        printer.print_list(self.offset, lambda x: printer.print_string(str(x)))
+        printer.print_string("] size [")
+        printer.print_list(self.size, lambda x: printer.print_string(str(x)))
+        printer.print_string("] source offset [")
+        printer.print_list(self.source_offset, lambda x: printer.print_string(str(x)))
+        printer.print_string("] to {}>".format(self.neighbor))
+
+    # TODO: def parse_parameters()
```

Comment:
can you file an issue if you're planning on adding it later? It feels doable as this PR also

Reply:
I'm not sure if we want to have issues for stuff in experimental, as that is more the "move fast and break things" part of the codebase and less of a managed process, so I feel having issues for everything that needs to be done in experimental is more of a hindrance than something we want...